File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    public static int[] SIGN = {1,-1};
    public static int ANSWER = 0;
    
    public int solution(int[] numbers, int target) {
        search(numbers, target, 0, 0, 0);
        search(numbers, target, 0, 1, 0);
        int answer = ANSWER;
        
        return answer;
    }
    
    public static void search(int[] numbers, int target, int numidx, int signidx, int sum){
        if(numidx+1 == numbers.length){
            sum = sum + (numbers[numidx] * SIGN[signidx]);
            if (sum == target){
                ANSWER++;
                return;
            }
            return;
        }
        else{
            sum = sum + (numbers[numidx] * SIGN[signidx]);
            search(numbers, target, numidx + 1, 0,sum);
            search(numbers, target, numidx + 1, 1,sum);
           }
    }
}

File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.py
import sys
from collections import deque

que = deque()
def bfs(numbers, target):
    ans = 0
    length = len(numbers)
    que.append((numbers[0], 0))
    que.append((-numbers[0], 0))
    while(len(que) > 0):
        total, depth = que.popleft()            
        if depth == length - 1:
            if total == target:
                ans += 1
        else:
            que.append((total + numbers[depth + 1], depth+1))
            que.append((total - numbers[depth + 1], depth+1))
    return ans
    
            

def solution(numbers, target):
    ans = bfs(numbers, target)
    return ans